docarray/doc_index/abstract_doc_index.py

```diff
@@ -626,31 +628,60 @@ def build_query(self) -> QueryBuilder:
         """
         return self.QueryBuilder()  # type: ignore
 
-    def _create_columns(self, schema: Type[BaseDocument]) -> Dict[str, _ColumnInfo]:
-        columns: Dict[str, _ColumnInfo] = dict()
+    def _flatten_schema(
+        self, schema: Type[BaseDocument], name_prefix: str = ''
+    ) -> List[Tuple[str, Type, 'ModelField']]:
+        """Flatten the schema of a Document into a list of column names and types.
+
+        :param schema: The schema to flatten
+        :param name_prefix: prefix to append to the column names. Used for recursive calls to handle nesting.
+        :return: A list of column names, types, and fields
+        """
+        names_types_fields: List[Tuple[str, Type, 'ModelField']] = []
         for field_name, field_ in schema.__fields__.items():
             t_ = schema._get_field_type(field_name)
+            inner_prefix = name_prefix + field_name + '__'
+
             if is_union_type(t_):
+                union_args = get_args(t_)
+                if len(union_args) == 2 and type(None) in union_args:
+                    # simple "Optional" type, treat as special case:
+                    # treat as if it was a single non-optional type
+                    for t_arg in union_args:
+                        if t_arg is type(None):
+                            pass
+                        elif issubclass(t_arg, BaseDocument):
```

Comment:
what if it is not a Union of document ? Example `Union[BaseDocument, int]` what should happen ? Failing ?

Reply:
it is just returned like that, without recursive flattening. I don't think we can do anything else. But I wouldn't fail here, since this only does the flattening as much as it can, and having this kind of thing returned might still be useful somewhere. But I guess it should fail at the call site